diff --git a/kernel/kthread.c b/kernel/kthread.c
index 933a625621b8..3fb1e7893dd7 100644
--- a/kernel/kthread.c
+++ b/kernel/kthread.c
@@ -11,6 +11,7 @@
 #include <linux/mm.h>
 #include <linux/mmu_context.h>
 #include <linux/sched.h>
+#include <linux/sched/isolation.h>
 #include <linux/sched/mm.h>
 #include <linux/sched/task.h>
 #include <linux/kthread.h>
diff --git a/kernel/sched/isolation.c b/kernel/sched/isolation.c
index 5a6ea03f9882..bd1ac3a37bf0 100644
--- a/kernel/sched/isolation.c
+++ b/kernel/sched/isolation.c
@@ -173,6 +173,24 @@ static int __init housekeeping_isolcpus_setup(char *str)
 			continue;
 		}
 
+		if (!strncmp(str, "wq,", 3)) {
+			str += 3;
+			flags |= HK_FLAG_WQ;
+			continue;
+		}
+
+		if (!strncmp(str, "rcu,", 4)) {
+			str += 4;
+			flags |= HK_FLAG_RCU;
+			continue;
+		}
+
+		if (!strncmp(str, "kthread,", 8)) {
+			str += 8;
+			flags |= HK_FLAG_KTHREAD;
+			continue;
+		}
+
 		/*
 		 * Skip unknown sub-parameter and validate that it is not
 		 * containing an invalid character.
